### PR TITLE
Autoproj

### PR DESCRIPTION
**Pull request by g-arjones**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -52,10 +132,25 @@
                 "category": "rock",
                 "command": "rock.selectPackage"
             },
+            {
+                "title": "Select package type",
```
Comment (doudou) [inline package.json]: Package types are listed in the installation manifest.

File: src/status.ts [hunk 2]
```diff
@@ -68,46 +81,96 @@ export class StatusBar implements vscode.Disposable {
         });
     }
 
-    public update() {
-        this.updateSelectedPackage();
-        this.updateBuildButton();
-        this.reloadVisibility();
+    public async update() {
+        await this.updateSelectedPackage();
```
Comment (doudou) [inline src/status.ts]: Each and every one of these re-resolves the "current package" wouldn't it make more sense to pass it as argument ?

**Comment by doudou** [comment]
Cool stuff.

I get `command 'rock.setDebuggingTarget' not found` as soon as I try to pick the `rock: set debugging target` command. Don't have the time to investigate further right now ... Will look at it tomorrow if you don't ;-)

File: package.json (hunk 1 above)
Comment (g-arjones) [reply package.json]: You probably noticed it already... This is supposed to be used as a fallback when the package is not registered in the manifest

File: src/status.ts (hunk 2 above)
Comment (g-arjones) [reply src/status.ts]: Definitely!

**Comment by g-arjones** [comment]
Probably a typo or something I renamed while refactoring... Should be in commands.Commands().register()

**Comment by g-arjones** [comment]
Did you take a look at [this](https://github.com/g-arjones/vscode-rock/blob/autoproj/src/async.ts#L99)? Thoughts? Pitfalls? I couldn't think of another way to run the debugger with the proper environment setup. It worked surprisingly well during my tests but I may be overlooking something...

**Comment by doudou** [comment]
> Did you take a look at this? Thoughts? Pitfalls?

I'm only starting to have a look .... wanted to try it out first to get the goodies ;-)

This is quite a bit of code, will take a bit of time :P

**Comment by doudou** [comment]
Hey. I see small changes that would IMO improve readability. I thought I would change those, and then send them to your for review/integration. Would that be OK ?

**Comment by g-arjones** [comment]
> Hey. I see small changes that would IMO improve readability. I thought I would change those, and then send them to your for review/integration. Would that be OK ?

Absolutely, yes! 👍

File: src/context.ts [hunk 3]
```diff
@@ -1,76 +1,154 @@
...
+        let data = this.loadPersistedData(path);
+        if (!data || !data.debuggingTarget ||
+            !data.debuggingTarget.name || !data.debuggingTarget.path)
+            return undefined;
+
+        return new debug.Target(data.debuggingTarget.name,
+            data.debuggingTarget.path);
+    }
+
+    public async getSelectedPackage(): Promise<packages.Package>
```
Comment (doudou) [inline src/context.ts]: I have the impression here that you assume that every editor/path is a package. Does that really make any sense at all ?

Comment (g-arjones) [reply src/context.ts]: I assume that every folder added to the vscode's workspace are either a configuration (autoproj/ and its childs) or a package. Then, getWorkspaceFolder(resource) gives me the parent folder (which I assume is a package). Thoughts?

**Comment by g-arjones** [comment]
> I get command 'rock.setDebuggingTarget' not found as soon as I try to pick the rock: set debugging target command.

Could not reproduce this one here.. Did you figure it out?

File: src/status.ts (hunk 2 above)
Comment (g-arjones) [reply src/status.ts]: Updated.

**Comment by doudou** [comment]
> Did you figure it out?

Not yet. I've been doing a lot of reading and some refactoring so far.

File: src/context.ts (hunk 3 above)
Comment (doudou) [reply src/context.ts]: > I assume that every folder added to the vscode's workspace are either a configuration (autoproj/ and its childs) or a package. 

I don't think it makes sense. People may have to add other packages outside of their workspace. I think it's taken care of in the refactoring I'm doing (through InvalidPackage).

Comment (g-arjones) [reply src/context.ts]: Current implementation allows packages from outside of an workspace. They are [ForeignPackages](https://github.com/g-arjones/vscode-rock/blob/autoproj/src/packages.ts#L362). Anyway, let's wait for your refactoring...

**Pull request merged**